interface/ceed-operator.c

```diff
@@ -248,6 +248,166 @@ int CeedCompositeOperatorAddSub(CeedOperator compositeop,
+
+  This returns a CeedVector containing a matrix at each quadrature point
+    providing the action of the CeedQFunction associated with the CeedOperator.
+    The vector 'assembled' is of shape
+      [num_elements, num_input_fields, num_output_fields, num_quad_points]
+    and contains column-major matrices representing the action of the
+    CeedQFunction for a corresponding quadrature point on an element. Inputs and
+    outputs are in the order provided by the user when adding CeedOperator fields.
+
+  @param op             CeedOperator to assemble CeedQFunction
+  @param[out] assembled CeedVector to store assembled Ceed QFunction at
+                          quadrature points
+  @param[out] rstr      CeedElemRestriction for CeedVector containing assembled
+                          CeedQFunction
+  @param request        Address of CeedRequest for non-blocking completion, else
+                          CEED_REQUEST_IMMEDIATE
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedOperatorAssembleLinearQFunction(CeedOperator op, CeedVector *assembled,
+    CeedElemRestriction *rstr, CeedRequest *request) {
+  int ierr;
+  Ceed ceed = op->ceed;
+  CeedQFunction qf = op->qf;
+
+  if (op->composite) {
+    return CeedError(ceed, 1, "Cannot assemble QFunction for composite operator");
+  } else {
+    if (op->nfields == 0)
+      return CeedError(ceed, 1, "No operator fields set");
+    if (op->nfields < qf->numinputfields + qf->numoutputfields)
+      return CeedError( ceed, 1, "Not all operator fields set");
+    if (op->numelements == 0)
+      return CeedError(ceed, 1, "At least one restriction required");
+    if (op->numqpoints == 0)
+      return CeedError(ceed, 1, "At least one non-collocated basis required");
+  }
+  ierr = op->AssembleLinearQFunction(op, assembled, rstr, request);
+  CeedChk(ierr);
+  return 0;
+}
+
+/**
+  @brief Build a FDM based approximate inverse for each element for a
+           CeedOperator
+
+  This returns a CeedOperator and CeedVector to apply a Fast Diagonalization
+    Method based approximate inverse. This function obtains the simultaneous
+    diagonalization for the 1D mass and Laplacian operators,
+      M = V^T V, K = V^T S V.
+    The assembled QFunction is used to modify the eigenvalues from simultaneous
+    diagonalization and obtain an approximate inverse of the form
+      V^T S^hat V.
+
+  @param op             CeedOperator to create element inverses
+  @param[out] fdminv    CeedOperator to apply the action of a FDM based inverse
+                          for each element
+  @param[out] qdata     CeedVector to hold qdata for fdminv
+  @param request        Address of CeedRequest for non-blocking completion, else
+                          CEED_REQUEST_IMMEDIATE
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Advanced
+**/
+int CeedOperatorCreateFDMElementInverse(CeedOperator op, CeedOperator *fdminv,
```

Comment:
@jedbrown, this is a WIP downstream of PR #306, but this is most of the code needed to provide per element inverses that we can (ideally) plug into @stefanozampini's branch. The exciting bit that will actually take some math is lines 374-377 below. I am going to start with a simple average of the QFunction values with qweight divided out and go from there.